### PR TITLE
gh-93466: Document PyType_Spec doesn't accept repeated slot IDs; raise where this was problematic

### PR DESCRIPTION
**Pull request by encukou**

### CODE REVIEW

**Comment by oda-gitso** [comment]
#93466 (so that viewers of the issue can see that a PR has been made).

File: Objects/typeobject.c [hunk 1]
```diff
@@ -3409,14 +3409,20 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
...
     char *res_start;
     short slot_offset, subslot_offset;
 
     nmembers = weaklistoffset = dictoffset = vectorcalloffset = 0;
     for (slot = spec->slots; slot->slot; slot++) {
         if (slot->slot == Py_tp_members) {
-            nmembers = 0;
+            if (nmembers != 0) {
+                PyErr_SetString(
+                    PyExc_RuntimeError,
```
Comment (erlend-aasland) [inline Objects/typeobject.c]: SystemError would be more appropriate. IIRC, that's what other similar error conditions raise.

File: Objects/typeobject.c [hunk 2]
```diff
@@ -3559,6 +3565,12 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
         else if (slot->slot == Py_tp_doc) {
             /* For the docstring slot, which usually points to a static string
                literal, we need to make a copy */
+            if (type->tp_doc != NULL) {
+                PyErr_SetString(
+                    PyExc_RuntimeError,
```
Comment (erlend-aasland) [inline Objects/typeobject.c]: Ditto

**Review by erlend-aasland — APPROVED** [review]
Looks good! I could not find other memcpy's, so this PR seems to be sufficient to prevent the leaks, however one might argue that we should raise errors for _any_ repeated slot.

**Comment by encukou** [comment]
Thanks for the review!

**Comment by sweeneyde** [comment]
I think this may have introduced a reference leak to `test_capi`. Maybe in the `"Multiple Py_tp_doc slots are not supported"` error case?

**Pull request merged**